Doxyfile

```diff
@@ -0,0 +1,2427 @@
+# Doxyfile 1.8.11
+
+# This file describes the settings to be used by the documentation system
+# doxygen (www.doxygen.org) for a project.
+#
+# All text after a double hash (##) is considered a comment and is placed in
+# front of the TAG it is preceding.
+#
+# All text after a single hash (#) is considered a comment and will be ignored.
+# The format is:
+# TAG = value [value, ...]
+# For lists, items can also be appended using:
+# TAG += value [value, ...]
+# Values that contain spaces should be placed between quotes (\" \").
+
+#---------------------------------------------------------------------------
+# Project related configuration options
+#---------------------------------------------------------------------------
+
+# This tag specifies the encoding used for all characters in the config file
+# that follow. The default is UTF-8 which is also the encoding used for all text
+# before the first occurrence of this tag. Doxygen uses libiconv (or the iconv
+# built into libc) for the transcoding. See http://www.gnu.org/software/libiconv
+# for the list of possible encodings.
+# The default value is: UTF-8.
+
+DOXYFILE_ENCODING      = UTF-8
+
+# The PROJECT_NAME tag is a single word (or a sequence of words surrounded by
+# double-quotes, unless you are using Doxywizard) that should identify the
+# project for which the documentation is generated. This name is used in the
+# title of most generated pages and in a few other places.
+# The default value is: My Project.
+
+PROJECT_NAME           = "Iroha/Hyperledger"
+
+# The PROJECT_NUMBER tag can be used to enter a project or revision number. This
+# could be handy for archiving the generated documentation or if some version
+# control system is used.
+
+PROJECT_NUMBER         =
+
+# Using the PROJECT_BRIEF tag one can provide an optional one line description
+# for a project that appears at the top of each page and should give viewer a
+# quick idea about the purpose of the project. Keep the description short.
+
+PROJECT_BRIEF          = "Iroha - A simple, decentralized ledger http://iroha.tech"
+
+# With the PROJECT_LOGO tag one can specify a logo or an icon that is included
+# in the documentation. The maximum height of the logo should not exceed 55
+# pixels and the maximum width should not exceed 200 pixels. Doxygen will copy
+# the logo to the output directory.
+
+PROJECT_LOGO           = ./docs/
+
+# The OUTPUT_DIRECTORY tag is used to specify the (relative or absolute) path
+# into which the generated documentation will be written. If a relative path is
+# entered, it will be relative to the location where doxygen was started. If
+# left blank the current directory will be used.
+
+OUTPUT_DIRECTORY       = ./docs/doxygen
```

Comment:
it generates build artifacts out of build directory, which is not good.

In CMake we specify build directory, from CMake it is possible to generate a documentation exactly in build directory.

In CMake 3.9 there is [FindDoxygen.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindDoxygen.cmake), which can be used to generate documentation without configuration file.

Since we require cmake minimal version 3.5.1 (and not 3.9), it is not guaranteed that this file will exist.
So, I propose you to: 
1. copy this file to `cmake/Modules`, 
2. create `cmake/documentation.cmake`, include there all config for docs (variables, function to generate docs, etc)
2. then in `cmake/dependencies.cmake` use `find_package(Doxygen)` and `include(cmake/documentation.cmake)`

If Doxygen is not found, do not interrupt configuration, just output something like "Doxygen is not found, docs target is disabled" (same as for cppcheck and gcovr targets)

Reply:
it will be discussed later on